vms/proposervm/proposer/windower.go

```diff
@@ -71,52 +108,32 @@ func New(state validators.State, subnetID, chainID ids.ID) Windower {
 }
 
 func (w *windower) Proposers(ctx context.Context, chainHeight, pChainHeight uint64, maxWindows int) ([]ids.NodeID, error) {
-	// get the validator set by the p-chain height
-	validatorsMap, err := w.state.GetValidatorSet(ctx, pChainHeight, w.subnetID)
+	validators, err := w.sortedValidators(ctx, pChainHeight)
 	if err != nil {
 		return nil, err
 	}
 
-	// convert the map of validators to a slice
-	validators := make([]validatorData, 0, len(validatorsMap))
-	weight := uint64(0)
-	for k, v := range validatorsMap {
-		validators = append(validators, validatorData{
-			id:     k,
-			weight: v.Weight,
-		})
-		newWeight, err := math.Add64(weight, v.Weight)
-		if err != nil {
-			return nil, err
-		}
-		weight = newWeight
-	}
-
-	// canonically sort validators
-	// Note: validators are sorted by ID, sorting by weight would not create a
-	// canonically sorted list
-	utils.Sort(validators)
-
-	// convert the slice of validators to a slice of weights
-	validatorWeights := make([]uint64, len(validators))
-	for i, v := range validators {
-		validatorWeights[i] = v.weight
-	}
-
-	seed := chainHeight ^ w.chainSource
+	var (
+		seed             = chainHeight ^ w.chainSource
+		source           = prng.NewMT19937()
+		validatorWeights = validatorsToWeight(validators)
+	)
 
-	source := prng.NewMT19937()
-	source.Seed(seed)
 	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
 	if err := sampler.Initialize(validatorWeights); err != nil {
 		return nil, err
 	}
 
-	numToSample := maxWindows
-	if weight < uint64(numToSample) {
-		numToSample = int(weight)
+	var totalWeight uint64
+	for _, weight := range validatorWeights {
+		totalWeight, err = math.Add64(totalWeight, weight)
+		if err != nil {
+			return nil, err
+		}
 	}
 
+	numToSample := int(math.Min(uint64(maxWindows), totalWeight))
```

Comment:
I'm not sure it is always safe to cast weight to int?

Reply:
I think this is safe. `maxWindows` is bounded to be smaller than `proposer.MaxBuildWindows`

Reply:
ah fair, duh.

Reply:
(and that is already an int)